services/src/main/java/io/grpc/protobuf/services/ProtoReflectionService.java

```diff
@@ -16,524 +16,100 @@
-          String serviceName = serviceDescriptor.getName();
-          FileDescriptor fileDescriptor =
-              ((ProtoFileDescriptorSupplier) serviceDescriptor.getSchemaDescriptor())
-                  .getFileDescriptor();
-          serverFileDescriptors.add(fileDescriptor);
-          serverServiceNames.add(serviceName);
-        }
-      }
-
-      // Replace the index if the underlying mutable services have changed. Check both the file
-      // descriptors and the service names, because one file descriptor can define multiple
-      // services.
-      FileDescriptorIndex mutableServicesIndex = index.getMutableServicesIndex();
-      if (!mutableServicesIndex.getServiceFileDescriptors().equals(serverFileDescriptors)
-          || !mutableServicesIndex.getServiceNames().equals(serverServiceNames)) {
-        index =
-            new ServerReflectionIndex(server.getImmutableServices(), serverMutableServices);
-        serverReflectionIndexes.put(server, index);
-      }
-
-      return index;
-    }
-  }
-
   @Override
-  public StreamObserver<ServerReflectionRequest> serverReflectionInfo(
-      final StreamObserver<ServerReflectionResponse> responseObserver) {
-    final ServerCallStreamObserver<ServerReflectionResponse> serverCallStreamObserver =
-        (ServerCallStreamObserver<ServerReflectionResponse>) responseObserver;
-    ProtoReflectionStreamObserver requestObserver =
-        new ProtoReflectionStreamObserver(getRefreshedIndex(), serverCallStreamObserver);
-    serverCallStreamObserver.setOnReadyHandler(requestObserver);
-    serverCallStreamObserver.disableAutoRequest();
-    serverCallStreamObserver.request(1);
-    return requestObserver;
+  public ServerServiceDefinition bindService() {
+    return createServerServiceDefinition();
   }
 
-  private static class ProtoReflectionStreamObserver
-      implements Runnable, StreamObserver<ServerReflectionRequest> {
-    private final ServerReflectionIndex serverReflectionIndex;
-    private final ServerCallStreamObserver<ServerReflectionResponse> serverCallStreamObserver;
-
-    private boolean closeAfterSend = false;
-    private ServerReflectionRequest request;
-
-    ProtoReflectionStreamObserver(
-        ServerReflectionIndex serverReflectionIndex,
-        ServerCallStreamObserver<ServerReflectionResponse> serverCallStreamObserver) {
-      this.serverReflectionIndex = serverReflectionIndex;
-      this.serverCallStreamObserver = checkNotNull(serverCallStreamObserver, "observer");
-    }
-
-    @Override
-    public void run() {
-      if (request != null) {
-        handleReflectionRequest();
-      }
-    }
+  private <ReqT, RespT> ServerServiceDefinition createServerServiceDefinition() {
+    ServerServiceDefinition serverServiceDefinitionV1 = ProtoReflectionServiceV1.newInstance()
+        .bindService();
+    MethodDescriptor<ReqT, RespT> methodDescriptorV1 = (MethodDescriptor<ReqT, RespT>) serverServiceDefinitionV1.getServiceDescriptor()
```

Comment:
After I changed the usage to ServerReflectionGrpc.getServerReflectionInfoMethod(); the generics problem is gone since this is a template specialized method with ServerReflectionRequest and ServerReflectionResponse in the method definition.